File: Data Structures & Algorithms/balanced-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if (root == null) return true;

        function height(root){
            if (root == null) return 0;

            let left = 1 + height(root.left);
            let right = 1 + height(root.right);
            if (right > left){
                return right;
            } else {
                return left;
            }
        }

        return (Math.abs(height(root.right) - height(root.left)) <= 1) && this.isBalanced(root.right) && this.isBalanced(root.left);

    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        r = []
        def dfs(root):
            if root is None:
                r.append("null")
                return
            r.append(str(root.val))
            dfs(root.left)
            dfs(root.right)
            return
        dfs(root)
        print(" ".join(r))
        return " ".join(r)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        root = data.split(" ")
        self.count = 0
        def dfs():
            if root[self.count] == "null":
                self.count += 1
                return None
            tree = TreeNode(root[self.count])
            self.count += 1
            tree.left = dfs()
            tree.right = dfs()
            return tree
        return dfs()
